nmpolicy/types.go

```diff
@@ -0,0 +1,51 @@
+/*
+ * This file is part of the nmpolicy project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package nmpolicy
+
+type PolicySpec struct {
+	Captures []CaptureSpec
+	DesiredState NMState
+}
+
+type CaptureSpec struct {
+	ID string
+	Expression string
+}
+
+type CacheState struct {
+	Captures []CaptureState
+}
+
+type GeneratedState struct {
+	Captures []CaptureState
```

Comment:
Since we don't care about the order of the captures, we can use a `map[string]string`.
We will need to have validation avoiding cyclic dependency and recursion (dependency to yourself).

Reply:
The OVS stuff is an example of a "surprise", something that was not expected.
When considering an API, I think it is not enough to examine what you know at the moment, but to be prepared for the future as well. This idea is not new, standards have been using this, I am not inventing anything.

The example with the nmstate spec is good IMO, it tracked a standard which placed the interfaces in a list and not a map, even though the interface name could be the only index. As we saw later, it is limiting, so having a list allowed indexing for more than a single field (e.g. name and type).

Anyway, this is just for the record, we have taken the approach of indexing only based on the name as suggested (i.e. a map).